File: CS/WinWebSolution.Module/AllowSortListViewController.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.Persistent.Base;
using DevExpress.Xpo;
using DevExpress.Persistent.BaseImpl;
using DevExpress.ExpressApp;
using DevExpress.Xpo.DB;
using DevExpress.Xpo.Metadata;
using DevExpress.ExpressApp.NodeWrappers;
using DevExpress.Data;
using DevExpress.ExpressApp.DC;
using DevExpress.ExpressApp.InfoGenerators;

namespace WinWebSolution.Module {
    [AttributeUsage(AttributeTargets.Interface | AttributeTargets.Class | AttributeTargets.Field | AttributeTargets.Property)]
    public class AllowSortAttribute : Attribute {
        public const string DefaultListViewAllowSortAttributeName = "DefaultListViewAllowSort";
        public const string AllowSortAttributeName = "AllowSort";
        public static AllowSortAttribute Default = new AllowSortAttribute(true);
        private bool allowSortCore = true;
        public AllowSortAttribute() : this(true) { }
        public AllowSortAttribute(bool allowSort) {
            this.allowSortCore = allowSort;
        }
        public bool AllowSort {
            get { return allowSortCore; }
            set { allowSortCore = value; }
        }
    }
    public abstract class AllowSortListViewController : ViewController<ListView>, IObjectModelCustomLoader {
        protected abstract void UpdateAllowSort();
        private void ListView_InfoChanged(object sender, EventArgs e) {
            UpdateAllowSort();
        }
        protected bool GetAllowSortListView() {
            if (View.Model != null) {
                return View.Model.Node.GetAttributeBoolValue(AllowSortAttribute.AllowSortAttributeName);
            }
            return false;
        }
        protected bool GetAllowSortColumn(ColumnInfoNodeWrapper columnInfo) {
            if (columnInfo != null && GetAllowSortListView()) {
                return columnInfo.Node.GetAttributeBoolValue(AllowSortAttribute.AllowSortAttributeName);
            }
            return false;
        }
        protected override void OnActivated() {
            base.OnActivated();
            View.InfoChanged += new EventHandler(ListView_InfoChanged);
        }
        protected override void OnDeactivating() {
            View.InfoChanged -= new EventHandler(ListView_InfoChanged);
            base.OnDeactivating();
        }
        void IObjectModelCustomLoader.CustomizeClassNode(ITypeInfo classInfo, ClassInfoNodeWrapper classNode) {
            AllowSortAttribute attr = classInfo.FindAttribute<AllowSortAttribute>() ?? AllowSortAttribute.Default;
            classNode.Node.SetAttribute(AllowSortAttribute.DefaultListViewAllowSortAttributeName, attr.AllowSort);
        }
        void IObjectModelCustomLoader.CustomizeMemberNode(IMemberInfo memberInfo, PropertyInfoNodeWrapper memberNode) {
            AllowSortAttribute attr = memberInfo.FindAttribute<AllowSortAttribute>() ?? AllowSortAttribute.Default;
            memberNode.Node.SetAttribute(AllowSortAttribute.AllowSortAttributeName, attr.AllowSort);
        }
        public override Schema GetSchema() {
            return new Schema(new DictionaryXmlReader().ReadFromString(
                @"<Element Name=""Application"">
                    <Element Name=""BOModel"">
                        <Element Name=""Class"">
                            <Attribute Name=""" + AllowSortAttribute.DefaultListViewAllowSortAttributeName + @""" Choice=""True,False"" />
                            <Element Name=""Member"">
								<Attribute Name=""" + AllowSortAttribute.AllowSortAttributeName + @""" Choice=""True,False""/>
							</Element>
                        </Element>
                    </Element>
                    <Element Name=""Views"">
		                <Element Name=""ListView"">
			                <Attribute Name=""" + AllowSortAttribute.AllowSortAttributeName + @"""
				                 DefaultValueExpr=""SourceNode=BOModel\Class\@Name=@ClassName; SourceAttribute=@" + AllowSortAttribute.DefaultListViewAllowSortAttributeName + @"""
			                     Choice=""True,False""
				                 />
                                <Element Name=""Columns"">
									<Element Name=""ColumnInfo"">
										<Attribute Name=""" + AllowSortAttribute.AllowSortAttributeName + @""" IsNewNode=""True"" 
											DefaultValueExpr=""{DevExpress.ExpressApp.Core.DictionaryHelpers.BOPropertyCalculator}ClassName=..\..\@ClassName""/>
									</Element>
								</Element>

		                </Element>
		            </Element>
                </Element>"));
        }
    }
}

File: VB/WinWebSolution.Module/AllowSortListViewController.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.Persistent.Base
Imports DevExpress.Xpo
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.ExpressApp
Imports DevExpress.Xpo.DB
Imports DevExpress.Xpo.Metadata
Imports DevExpress.ExpressApp.NodeWrappers
Imports DevExpress.Data
Imports DevExpress.ExpressApp.DC
Imports DevExpress.ExpressApp.InfoGenerators
Imports DevExpress.ExpressApp.Model

Namespace WinWebSolution.Module
	<AttributeUsage(AttributeTargets.Interface Or AttributeTargets.Class Or AttributeTargets.Field Or AttributeTargets.Property)> _
	Public Class AllowSortAttribute
		Inherits Attribute
		Public Const DefaultListViewAllowSortAttributeName As String = "DefaultListViewAllowSort"
		Public Const AllowSortAttributeName As String = "AllowSort"
		Public Shared [Default] As New AllowSortAttribute(True)
		Private allowSortCore As Boolean = True
		Public Sub New()
			Me.New(True)
		End Sub
		Public Sub New(ByVal allowSort As Boolean)
			Me.allowSortCore = allowSort
		End Sub
		Public Property AllowSort() As Boolean
			Get
				Return allowSortCore
			End Get
			Set(ByVal value As Boolean)
				allowSortCore = value
			End Set
		End Property
	End Class
	Public MustInherit Class AllowSortListViewController
		Inherits ViewController(Of ListView)
		Protected MustOverride Sub UpdateAllowSort()
		Private Sub ListView_InfoChanged(ByVal sender As Object, ByVal e As EventArgs)
			UpdateAllowSort()
		End Sub
		Protected Function GetAllowSortListView() As Boolean
			If View.Model IsNot Nothing Then
				Return (CType(View.Model, IModelListViewAllowSort)).AllowSort
			End If
			Return False
		End Function
		Protected Function GetAllowSortColumn(ByVal column As IModelColumn) As Boolean
			If column IsNot Nothing AndAlso GetAllowSortListView() Then
				Return (CType(column, IModelColumnAllowSort)).AllowSort
			End If
			Return False
		End Function
		Protected Overrides Overloads Sub OnActivated()
			MyBase.OnActivated()
			AddHandler View.ModelChanged, AddressOf ListView_InfoChanged
		End Sub
		Protected Overrides Overloads Sub OnDeactivated()
			RemoveHandler View.ModelChanged, AddressOf ListView_InfoChanged
			MyBase.OnDeactivated()
		End Sub
		'void IObjectModelCustomLoader.CustomizeClassNode(ITypeInfo classInfo, ClassInfoNodeWrapper classNode) {
		'    AllowSortAttribute attr = classInfo.FindAttribute<AllowSortAttribute>() ?? AllowSortAttribute.Default;
		'    classNode.Node.SetAttribute(AllowSortAttribute.DefaultListViewAllowSortAttributeName, attr.AllowSort);
		'}
		'void IObjectModelCustomLoader.CustomizeMemberNode(IMemberInfo memberInfo, PropertyInfoNodeWrapper memberNode) {
		'    AllowSortAttribute attr = memberInfo.FindAttribute<AllowSortAttribute>() ?? AllowSortAttribute.Default;
		'    memberNode.Node.SetAttribute(AllowSortAttribute.AllowSortAttributeName, attr.AllowSort);
		'}
	   ' public override Schema GetSchema() {
	   '     return new Schema(new DictionaryXmlReader().ReadFromString(
	   '         @"<Element Name=""Application"">
	   '             <Element Name=""BOModel"">
	   '                 <Element Name=""Class"">
	   '                     <Attribute Name=""" + AllowSortAttribute.DefaultListViewAllowSortAttributeName + @""" Choice=""True,False"" />
	   '                     <Element Name=""Member"">
							'	<Attribute Name=""" + AllowSortAttribute.AllowSortAttributeName + @""" Choice=""True,False""/>
							'</Element>
	   '                 </Element>
	   '             </Element>
	   '             <Element Name=""Views"">
			 '           <Element Name=""ListView"">
				'            <Attribute Name=""" + AllowSortAttribute.AllowSortAttributeName + @"""
				   '              DefaultValueExpr=""SourceNode=BOModel\Class\@Name=@ClassName; SourceAttribute=@" + AllowSortAttribute.DefaultListViewAllowSortAttributeName + @"""
				'                 Choice=""True,False""
				   '              />
	   '                         <Element Name=""Columns"">
							'		<Element Name=""ColumnInfo"">
							'			<Attribute Name=""" + AllowSortAttribute.AllowSortAttributeName + @""" IsNewNode=""True"" 
							'				DefaultValueExpr=""{DevExpress.ExpressApp.Core.DictionaryHelpers.BOPropertyCalculator}ClassName=..\..\@ClassName""/>
							'		</Element>
							'	</Element>

			 '           </Element>
			 '       </Element>
	   '         </Element>"));
	   ' }
	End Class
End Namespace